File: langs/typescript/src/collections/vmCollection.ts
import type { Observable } from "rxjs";
import type { ComponentVMBase } from "../components/componentVMBase.js";
import type { BatchUpdateHandle } from "./batchUpdateHandle.js";
import type { CollectionChangedEvent } from "./collectionChangedEvent.js";

/** Shared ordered, observable child-collection capability without selection. */
export interface IVmCollection<VM extends ComponentVMBase> extends Iterable<VM> {
  readonly count: number;
  readonly collectionChanged: Observable<CollectionChangedEvent>;
  at(index: number): VM;
  add(item: VM): void;
  insert(index: number, item: VM): void;
  remove(item: VM): boolean;
  removeAt(index: number): void;
  setAt(index: number, value: VM): void;
  clear(): void;
  move(fromIndex: number, toIndex: number): void;
  batchUpdate(): BatchUpdateHandle;
}

/** VM collection that additionally owns a current-child selection slot. */
export interface ISelectableVmCollection<VM extends ComponentVMBase>
  extends IVmCollection<VM> {
  current: VM | null;
  selectComponent(vm: VM): void;
  deselectComponent(vm: VM): void;
  canSelectComponent(vm: VM): boolean;
}

File: langs/swift/Sources/VMx/Collections/VMCollection.swift
import Combine

/// Catchable bounds failure for `VMCollection.move(from:to:)`.
public struct VMCollectionIndexError: Error, Equatable {
    public let index: Int
    public let count: Int

    public init(index: Int, count: Int) {
        self.index = index
        self.count = count
    }
}

/// Shared ordered, observable child-collection capability without selection.
public protocol VMCollection: AnyObject, Sequence where Element == Child {
    associatedtype Child: ComponentVMBase

    var count: Int { get }
    var collectionChanged: AnyPublisher<CollectionChangedEvent, Never> { get }
    func at(_ index: Int) -> Child
    func add(_ child: Child)
    func insert(_ child: Child, at index: Int)
    func remove(_ child: Child) -> Bool
    func removeAt(_ index: Int)
    func replace(at index: Int, with child: Child)
    func clear()
    func move(from fromIndex: Int, to toIndex: Int) throws
    func batchUpdate() -> BatchUpdateHandle
}

/// VM collection that additionally owns a current-child selection slot.
public protocol SelectableVMCollection: VMCollection {
    var current: Child? { get set }
    func selectComponent(_ vm: Child) throws
    func deselectComponent(_ vm: Child) throws
    func canSelectComponent(_ vm: Child) -> Bool
}
